pkg/nodeagent/controller/operatingsystemconfig/reconciler.go

```diff
@@ -0,0 +1,318 @@
+			if err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+				return fmt.Errorf("unable to read existing unit file %q for %q: %w", unitFilePath, unit.Name, err)
+			}
+
+			newUnitContent := []byte(*unit.Content)
+			if !bytes.Equal(newUnitContent, oldUnitContent) {
+				if err := afero.WriteFile(r.FS, unitFilePath, newUnitContent, 0600); err != nil {
+					return fmt.Errorf("unable to write unit file %q for %q: %w", unitFilePath, unit.Name, err)
+				}
+				log.Info("Successfully applied new or changed unit file", "path", unitFilePath)
+			}
+		}
+
+		for _, dropIn := range unit.DropIns {
+			dropInFilePath := path.Join(unitFilePath+".d", dropIn.Name)
+
+			oldDropInContent, err := afero.ReadFile(r.FS, dropInFilePath)
+			if err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+				return fmt.Errorf("unable to read existing drop-in file %q for unit %q: %w", dropInFilePath, unit.Name, err)
+			}
+
+			newDropInContent := []byte(dropIn.Content)
+			if !bytes.Equal(newDropInContent, oldDropInContent) {
+				if err := afero.WriteFile(r.FS, dropInFilePath, newDropInContent, 0600); err != nil {
+					return fmt.Errorf("unable to write drop-in file %q for unit %q: %w", dropInFilePath, unit.Name, err)
+				}
+				log.Info("Successfully applied new or changed drop-in file for unit", "path", dropInFilePath, "unit", unit.Name)
+			}
+		}
+
+		if pointer.BoolDeref(unit.Enable, true) {
+			if err := r.DBus.Enable(ctx, unit.Name); err != nil {
+				return fmt.Errorf("unable to enable unit %q: %w", unit.Name, err)
+			}
+			log.Info("Successfully enabled unit", "unitName", unit.Name)
+		} else {
+			if err := r.DBus.Disable(ctx, unit.Name); err != nil {
+				return fmt.Errorf("unable to disable unit %q: %w", unit.Name, err)
+			}
+			log.Info("Successfully disabled unit", "unitName", unit.Name)
+		}
+	}
+
+	return nil
+}
+
+func (r *Reconciler) removeDeletedUnits(ctx context.Context, log logr.Logger, node client.Object, units []extensionsv1alpha1.Unit) error {
+	for _, unit := range units {
+		if err := r.DBus.Stop(ctx, r.Recorder, node, unit.Name); err != nil {
+			return fmt.Errorf("unable to stop deleted unit %q: %w", unit.Name, err)
+		}
+
+		if err := r.DBus.Disable(ctx, unit.Name); err != nil {
+			return fmt.Errorf("unable to disable deleted unit %q: %w", unit.Name, err)
+		}
+
+		if err := r.FS.Remove(path.Join(etcdSystemdSystem, unit.Name)); err != nil && !errors.Is(err, afero.ErrFileNotFound) {
+			return fmt.Errorf("unable to delete systemd unit of deleted %q: %w", unit.Name, err)
+		}
+
+		log.Info("Successfully removed no longer needed unit", "unitName", unit.Name)
+	}
+
+	return nil
+}
+
+func (r *Reconciler) executeUnitCommands(ctx context.Context, log logr.Logger, node client.Object, units []changedUnit) error {
+	var fns []flow.TaskFn
+
+	for _, u := range units {
+		unit := u
+
+		command := extensionsv1alpha1.CommandRestart
+		if unit.Command != nil {
+			command = *unit.Command
+		}
```

Comment:
When a unit becomes disabled it should be stopped here, shouldn't it?

Reply:
You are also right